src/main/java/net/opentsdb/tsd/KafkaRpcPluginThread.java

```diff
@@ -240,114 +233,110 @@ Properties buildConsumerProperties() {
+                final double waiting = rate_limiter.acquire();
+                cumulativeRateDelay.addAndGet(waiting);
+                datapointsReceived.addAndGet(eventList.size());
+                for (TypedIncomingData ev : eventList) {
+                  ev.processData(this, recvTime);
+                }
+                break;
+              case REQUEUE_RAW:
+              case REQUEUE_ROLLUP:
+              case UID_ABUSE:
+                final List<TypedIncomingData> requeuedList =
+                  deserializer.deserialize(this, record.value().get());
+                if (requeuedList == null) {
+                  deserializationErrors.incrementAndGet();
+                  continue;
+                }
+
+                // to avoid tight requeue loops we want to sleep a spell
+                // if we receive a data point that was recently added
+                // to the queue
+                datapointsReceived.addAndGet(requeuedList.size());
+                for (TypedIncomingData ev : requeuedList) {
+                  final long requeueDiff = System.currentTimeMillis() - ev.getRequeueTS();
+                  if (requeueDiff < requeue_delay) {
+                    if (LOG.isDebugEnabled()) {
+                      LOG.debug("Sleeping for " + (requeue_delay - requeueDiff)
+                        + " ms due to requeue delay");
+                    }
+                    //incrementCounter(CounterType.RequeuesDelayed, ns);
+                    Thread.sleep(requeue_delay - requeueDiff);
+                  }
+
+                  // I &#9825 Google! It's so easy! No release necessary! Thread Safe!
+                  final double w = rate_limiter.acquire();
+                  cumulativeRateDelay.addAndGet(w);
+                  ev.processData(this, recvTime);
+                }
+                break;
+              default:
+                throw new IllegalStateException("Unknown consumer type: " +
                   consumer_type + " for " + this);
+            }
+            errorCount = 0;
+          } catch (RuntimeException e) {
+            LOG.error("Exception in kafkaReader or Tsdb Writer ", e);
+            incrementNamespaceCounter(CounterType.Exception, null);
+            errorCount++;
+            if (errorCount >= 5) {
+              LOG.error("Too many errors, Killing the consumer thread " + this);
+              throw e;
+            }
           }
-          errorCount = 0;
-        } catch (RuntimeException e) {
-          LOG.error("Exception in kafkaReader or Tsdb Writer ", e);
-          incrementNamespaceCounter(CounterType.Exception, null);
-          errorCount++;
-          if (errorCount >= 5) {
-            LOG.error("Too many errors, Killing the consumer thread " + this);
-            throw e;
-          }
+          nanoCtr = System.nanoTime();
         }
-        nanoCtr = System.nanoTime();
+        consumer.commitSync();
```

Comment:
Offset commiting is here

Reply:
Fwiw, we cherry-picked these changes to the version we're runing and have found that committing manually is significantly reducing the throughput. The synchronous call was the bottleneck and since we prioritise latency, we switched back to using auto-commits. It's not a problem if a couple of points are written multiple times if a commit fails as long as these failres are rare.

In our setup, the node we ran with this version of the plugin managed to write up to 9,000 datapoints per second and after switching to auto-commits every 5000 ms, we reached more than 40,000.

Whether it's necessary to commit after every poll is maybe something to reconsider for this PR. As a middle way, we have seen good results with making the commit call async.

Reply:
Yeah I was concerned about data durability, hence doing the commits manually and waiting for them to finish. However, if submitting the same metric multiple times just overwrites the same data, it shouldn't be a problem.